src/hotspot/cpu/x86/c2_stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,1850 @@
+  //  For UL where the needle size is > MAX_NEEDLE_LEN_TO_EXPAND and the haystack size minus
+  //  the needle size is less than 32 bytes, we default to a
+  //  byte-by-byte comparison (this will be rare).
+  //
+  //  Note that the code assumes MAX_NEEDLE_LEN_TO_EXPAND is >= 32.
+  //
+  //  The UU and LL cases are identical except for the loop increments and loading
+  //  of the characters into registers.  UU loads and compares words, LL - bytes.
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+
+  const Register haystack_p     = c_rarg0;
+  const Register haystack_len_p = c_rarg1;
+  const Register needle_p       = c_rarg2;
+  const Register needle_len_p   = c_rarg3;
+
+  // Addresses of the two jump tables used for small needle processing
+  address big_jump_table;
+  address small_jump_table;
+
+  Label L_begin;
+
+  Label L_returnRBP, L_checkRangeAndReturn, L_returnError, L_bigCaseFixupAndReturn;
+  Label L_bigSwitchTop, L_bigCaseDefault, L_smallCaseDefault;
+  Label L_nextCheck, L_checksPassed, L_zeroCheckFailed, L_return;
+  Label L_wcharBegin, L_continue, L_wideNoExpand;
+
+  __ align(CodeEntryAlignment);
+  fnptrs[ae] = __ pc();
+  __ enter();  // required for proper stackwalking of RuntimeStub frame
+
+  // Check for trivial cases
+  // needle length == 0?
+  __ cmpq(needle_len_p, 0);
+  __ jg_b(L_nextCheck);
+  __ xorl(rax, rax);
+  __ leave();
+  __ ret(0);
+
+  // haystack length == 0?
+  __ bind(L_nextCheck);
+  __ testq(haystack_len_p, haystack_len_p);
+  __ je(L_zeroCheckFailed);
+
+  // haystack length >= needle length?
+  __ movq(rax, haystack_len_p);
+  __ subq(rax, needle_len_p);
+  __ jge_b(L_checksPassed);
+
+  __ bind(L_zeroCheckFailed);
+  __ movq(rax, -1);
+  __ leave();
+  __ ret(0);
+
+  __ bind(L_checksPassed);
+
+  // Check for highly-optimized ability - haystack <= 32 bytes and needle <= 6 bytes
+  // haystack_len is in elements, not bytes, for UTF-16
+  __ cmpq(haystack_len_p, isU ? 16 : 32);
+  __ ja(L_begin);
+
+  // needle_len is in elements, not bytes, for UTF-16 <=> UTF-16
+  __ cmpq(needle_len_p, isUU ? 3 : 6);
+  __ ja(L_begin);
+
+  // Handle short haystack and needle specially
+  // Generated code does not return - either found or not
+  highly_optimized_short_cases(ae, haystack_p, haystack_len_p, needle_p, needle_len_p, xmm0, xmm1,
+                               r10, r11, _masm);
+
+  // If we're generating UL, we need to "pretend" we're generating UU code
+  // for the case where the needle can be expanded onto the stack
+  if (isReallyUL) {
+    ae = StrIntrinsicNode::UU;
+    isUL = false;
+    isUU = true;
+  }
+
+  // Set up jump tables.  Used when needle size <= NUMBER_OF_CASES
+  setup_jump_tables(ae, L_returnRBP, L_checkRangeAndReturn, L_bigCaseFixupAndReturn,
+                    &big_jump_table, &small_jump_table, _masm);
```

Comment:
We could directly use L_returnError here instead of L_returnRBP.

Reply:
OK